Remove duplicated function calls in generated shader

glfmn opened the pull request:
Remove re-calculation of UV coordinates in glsl

Alter the shader generator to assign to variables instead of nesting expressions, as some expressions get copied into multiple places when using any function that takes generators as inputs.

Consider the example:

```js
osc().modulate(voronoi()).rotate().scale().out()
```

This will generate the follwing GLSL (formatted for legibility)

```glsl
osc(
  modulate(
    rotate(scale(st, 1.5, 1., 1., 0.5, 0.5), 10., 0.),
    voronoi(
      rotate(scale(st, 1.5, 1., 1., 0.5, 0.5), 10., 0.)
      , 5., 0.3, 0.3)
    , 0.1)
  , 60., 0.1, 0.
);
```

The `rotate(scale(st, 1.5, 1., 1., 0.5, 0.5), 10., 0.)` portion nests into all generators before the `rotate` and `scale` calls in the original javascript, causing repeated calculations.

This change uses variable assignments to compose the hydra instead. The same example above outputs the glsl:

```glsl
st = scale(st, 1.5, 1., 1., 0.5, 0.5);
st = rotate(st, 10., 0.);
vec2 st_1 = st;
// modulate inputs:
vec4 c_m1 = voronoi(st_1, 5., 0.3, 0.3);
st = modulate(st, c_m1, 0.1);
vec4 c = osc(st, 60., 0.1, 0.);
```

The calls to `scale`, and `rotate` alter the uv coordinates directly; when a recursive call occurs in the shader generation, the generator creates new variables to ensure changes to the uv coordinates only effect the intended scope.

Tested with every example that I could get from `hydra.ojack.xyz` to ensure the generator produces functionally identical output.

![Screenshot from 2024-11-14 18-31-44](https://github.com/user-attachments/assets/3c10d7e6-2ee8-4a48-adc9-57d0c87def2e)

Commit:
Push extra inputs for types with more than 1 input

Modify `typeLookup` to store type and name information so that
`processGlsl` does not need to manually add extra inputs for function
signatures with more than 1 special input.

Addresses the following todo:

> add extra input to beginning for backward combatibility
> @todo update compiler so this is no longer necessary

Commit:
Remove re-calculation of UV coordinates in glsl

Alter the shader generator to assign to variables instead of nesting
expressions, as some expressions get copied into multiple places when
using any function that takes generators as inputs.

Consider the example:

```js
osc().modulate(voronoi()).rotate().scale().out()
```

This will generate the follwing GLSL (formatted for legibility)

```glsl
osc(
  modulate(
    rotate(scale(st, 1.5, 1., 1., 0.5, 0.5), 10., 0.),
    voronoi(
      rotate(scale(st, 1.5, 1., 1., 0.5, 0.5), 10., 0.)
      , 5., 0.3, 0.3)
    , 0.1)
  , 60., 0.1, 0.
);
```

The `rotate(scale(st, 1.5, 1., 1., 0.5, 0.5), 10., 0.)` portion nests
into all generators before the `rotate` and `scale` calls in the
original javascript, causing repeated calculations.

This change uses variable assignments to compose the hydra instead. The
same example above outputs the glsl:

```glsl
st = scale(st, 1.5, 1., 1., 0.5, 0.5);
st = rotate(st, 10., 0.);
vec2 st_1 = st;
// modulate inputs:
vec4 c_m1 = voronoi(st_1, 5., 0.3, 0.3);
st = modulate(st, c_m1, 0.1);
vec4 c = osc(st, 60., 0.1, 0.);
```

The calls to `scale`, and `rotate` alter the uv coordinates
directly; when a recursive call occurs in the shader generation, the
generator creates new variables to ensure changes to the uv coordinates
only effect the intended scope.

Tested with every example that I could get from `hydra.ojack.xyz` to
ensure the generator produces functionally identical output.

glfmn commented:
Hmm seem to have the wrong base commit for this branch, will rebase.

geikha commented:
Omg this is beautiful! thanks so much for your work! this is also a long awaited one no one had had the time (or clarity) to work on!

I have many ideas for the parsing I'd love to discuss and work together, is there anywhere I can reach you for starters? Of course most of the discussion is better to have publicly but just to get to know each other if you're up to it 👍

glfmn commented:
Glad to be of service,  I’d love to be more involved. You can email “me [at] glfmn [dot] io” if you’d like and we can arrange some way to meet up there!

glfmn commented on src/generate-glsl.js:
`shaderString` calls `generateGlsl` internally with the wrong patameters and need to be updated.

glfmn replied:
[9c32cae](https://github.com/hydra-synth/hydra-synth/pull/173/commits/9c32caeeb3e8c28875a41595d8ffbb8c4f7fde6c)

Commit:
Generalize input generation

Rather than specially process the transform-type based arguments, use
new function `generateInputs` to generate variables for every input
which is a generator/transform.

Then, in the `shaderString` function, use the variable that would be
generated by `generateInputs` as the input. An improvement would be
to make a new function which generates the variable name that can be
re-used, in order to make the dependency between these two functions
more obvious.

This renders a few different types semi-obsolete:

- `combine`: the code is exactly the same as for `color`
- `combineCoord`: the code is exactly the same as for `coord`

This makes sense, as for the purposes of chaining transforms together,
there are only 3 shapes or categories of functions:

- Functions which start the chain and produce the color var
- functions which assign to the color
- functions which assign to the uv

All inputs follow the same pattern otherwise, and it is the function
body (which is irrelevant to the shader generator) which gives those
inputs meaning.

I believe the `combine` type is still very useful for user discovery ;
however, while `combineCoord` provides similar utility, I believe it
is a less essential category than `combine`.

Commit:
Add `generateInputName` function

Commit:
Remove generated comments

Commit:
Remove unused shaderParams arg

Commit:
Further simplify fn input handling in processGlsl

We can address the previous @todo in the same way we addressed the
second todo.

Commit:
Add comment explaining why we slice input by 1
